flink-staging/flink-streaming/flink-streaming-core/src/test/java/org/apache/flink/streaming/api/functions/sink/SocketClientSinkTest.java

```diff
@@ -0,0 +1,301 @@
+import org.apache.flink.streaming.util.serialization.SerializationSchema;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import java.net.ServerSocket;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Tests for the {@link org.apache.flink.streaming.api.functions.sink.SocketClientSink}.
+ */
+public class SocketClientSinkTest{
+
+	final AtomicReference<Throwable> error = new AtomicReference<Throwable>();
+	private final String host = "127.0.0.1";
+	private int port;
+	private String value;
+
+	public Thread t;
+
+	public SocketClientSinkTest() {
+	}
+
+	@Test
+	public void testSocketSink() throws Exception{
+		value = "";
+		ServerSocket server = new ServerSocket(0);
+		port = server.getLocalPort();
+
+		new Thread(new Runnable() {
+			@Override
+			public void run() {
+				t = Thread.currentThread();
+				SerializationSchema<String, byte[]> simpleSchema = new SerializationSchema<String, byte[]>() {
+					@Override
+					public byte[] serialize(String element) {
+						return element.getBytes();
+					}
+				};
+
+				try {
+					SocketClientSink<String> simpleSink = new SocketClientSink<String>(host, port, simpleSchema, 0);
+					simpleSink.open(new Configuration());
+					simpleSink.invoke("testSocketSinkInvoke");
+					simpleSink.close();
+				} catch (Exception e){
+					error.set(e);
+				}
+			}
+		}).start();
+
+		Socket sk = server.accept();
+		BufferedReader rdr = new BufferedReader(new InputStreamReader(sk
+				.getInputStream()));
+		value = rdr.readLine();
+
+		t.join();
+		server.close();
+
+		if (error.get() != null) {
+			Throwable t = error.get();
+			t.printStackTrace();
+			fail("Error in spawned thread: " + t.getMessage());
+		}
+
+		assertEquals("testSocketSinkInvoke", value);
+	}
+
+	@Test
+	public void testSocketSinkNoRetry() throws Exception{
+		ServerSocket server = new ServerSocket(0);
+		port = server.getLocalPort();
+
+		new Thread(new Runnable() {
+
+			@Override
+			public void run() {
+				t = Thread.currentThread();
+				SerializationSchema<String, byte[]> simpleSchema = new SerializationSchema<String, byte[]>() {
+					@Override
+					public byte[] serialize(String element) {
+						return element.getBytes();
+					}
+				};
+
+				try {
+					SocketClientSink<String> simpleSink = new SocketClientSink<String>(host, port, simpleSchema, 0);
+					simpleSink.open(new Configuration());
+
+					synchronized (t) {
+						//wating for server to close
+						t.wait();
+					}
+
+					//firstly send a fin to cancel the client state:FIN_WAIT_2 while the server is CLOSE_WAIT
+					simpleSink.invoke("testSocketSinkInvoke");
+
+					//socket is closed then test "retry"
+					simpleSink.invoke("testSocketSinkInvoke");
+					simpleSink.close();
+					assertEquals(0, simpleSink.retries);
+				} catch (Exception e) {
+					//This Exception is good since the socket server is never reopen.
```

Comment:
Since the socket server never reopen.
The reconnecting will throw a exception when finished to retry.
Originate:(SocketClientSink.java:127)
if (!success) {
				throw new RuntimeException("Cannot send message " + value.toString() +
						" to socket server at " + hostName + ":" + port, e);
			}